gusto/time_discretisation.py

```diff
@@ -1084,3 +1090,463 @@ def apply(self, x_out, *x_in):
+            all_terms,
+            map_if_true=replace_subject(self.x1, self.idx))
+        r = r.label_map(lambda t: t.has_label(time_derivative),
+                        map_if_false=lambda t: -0.5*self.dt*t)
+
+        return r.form
+
+    @property
+    def lhs0(self):
+        """Set up the time discretisation's right hand side."""
+        l = self.residual.label_map(
+            all_terms,
+            map_if_true=replace_subject(self.x_out, self.idx))
+        l = l.label_map(lambda t: t.has_label(time_derivative),
+                        map_if_false=lambda t: 0.5*self.dt*t)
+        return l.form
+
+    @property
+    def lhs(self):
+        """Set up the time discretisation's right hand side."""
+        l = self.residual.label_map(
+            all_terms,
+            map_if_true=replace_subject(self.x_out, self.idx))
+        l = l.label_map(lambda t: t.has_label(time_derivative),
+                        map_if_false=lambda t: (5/12.)*self.dt*t)
+        return l.form
+
+    @property
+    def rhs(self):
+        """Set up the discretisation's right hand side."""
+        rn = self.residual.label_map(all_terms,
+                                     map_if_true=replace_subject(self.x1, self.idx))
+        rn = rn.label_map(lambda t: t.has_label(time_derivative),
+                          map_if_false=lambda t: -(8/12.)*self.dt*t)
+        rnm1 = self.residual.label_map(lambda t: t.has_label(time_derivative),
+                                       map_if_true=drop,
+                                       map_if_false=replace_subject(self.xnm1, self.idx))
+        rnm1 = rnm1.label_map(all_terms,
+                              map_if_true=lambda t: (1/12.)*self.dt*t)
+        r = rn + rnm1
+        return r.form
+
+    @property
+    def solver0(self):
+        """Set up the problem and the solver."""
+        # setup solver using lhs and rhs defined in derived class
+        problem = NonlinearVariationalProblem(self.lhs0-self.rhs0, self.x_out, bcs=self.bcs)
+        solver_name = self.field_name+self.__class__.__name__+"0"
+        return NonlinearVariationalSolver(problem, solver_parameters=self.solver_parameters, options_prefix=solver_name)
+
+    @property
+    def solver(self):
+        """Set up the problem and the solver."""
+        # setup solver using lhs and rhs defined in derived class
+        problem = NonlinearVariationalProblem(self.lhs-self.rhs, self.x_out, bcs=self.bcs)
+        solver_name = self.field_name+self.__class__.__name__
+        return NonlinearVariationalSolver(problem, solver_parameters=self.solver_parameters, options_prefix=solver_name)
+
+    def apply(self, x_out, *x_in):
+        """
+        Apply the time discretisation to advance one whole time step.
+
+        Args:
+            x_out (:class:`Function`): the output field to be computed.
+            x_in (:class:`Function`): the input field(s).
+        """
+        if self.initial_timesteps < self.nlevels-1:
```

Comment:
We should make an issue about how to deal with these time steppers when checkpointing

Reply:
Yep agreed! I have made one